genie-web/src/main/java/com/netflix/genie/web/services/loadbalancers/script/ScriptLoadBalancer.java

```diff
@@ -174,139 +236,139 @@ public Cluster selectCluster(
-                            tags.add(Tag.of(MetricsConstants.TagKeys.STATUS, STATUS_TAG_FOUND));
-                            return cluster;
-                        }
-                    }
-                    log.warn("Script returned a cluster not in the input list: {}", clusterId);
-                    tags.add(Tag.of(MetricsConstants.TagKeys.STATUS, STATUS_TAG_NOT_FOUND));
-                    return null; // Should throw
-                } else {
-                    tags.add(Tag.of(MetricsConstants.TagKeys.STATUS, STATUS_TAG_NO_PREFERENCE));
-                    log.debug("Script returned null, no preference");
-                    return null;
-                }
-            } else {
-                log.debug("Script not configured");
-                tags.add(Tag.of(MetricsConstants.TagKeys.STATUS, STATUS_TAG_NOT_CONFIGURED));
-                return null; // Should throw
-            }
-        } catch (final Exception e) {
-            tags.add(Tag.of(MetricsConstants.TagKeys.STATUS, STATUS_TAG_FAILED));
-            tags.add(Tag.of(MetricsConstants.TagKeys.EXCEPTION_CLASS, e.getClass().getCanonicalName()));
-            log.error("Unable to execute script due to {}", e.getMessage(), e);
-            return null; // Should throw
-        } finally {
-            this.registry
-                .timer(SELECT_TIMER_NAME, tags)
-                .record(System.nanoTime() - selectStart, TimeUnit.NANOSECONDS);
+                    .submit(() -> (String) script.eval(bindings))
+                    .get(timeoutLength, TimeUnit.MILLISECONDS);
+
+                log.debug("Script evaluated with result: {}", clusterId);
+
+                return clusterId;
         }
     }
 
-    /**
-     * Check if the script file needs to be refreshed.
-     */
-    public void refresh() {
-        log.debug("Refreshing");
-        final long updateStart = System.nanoTime();
-        final Set<Tag> tags = Sets.newHashSet();
-        try {
-            this.isUpdating.set(true);
+    static class Loader {
+        private final AtomicBoolean isUpdating = new AtomicBoolean();
+        private final AtomicBoolean isConfigured = new AtomicBoolean();
+        private final AtomicReference<CompiledScript> script = new AtomicReference<>(null);
+        private final ScriptEngineManager scriptEngineManager = new ScriptEngineManager();
 
-            // Update the script timeout
-            this.timeoutLength.set(
-                this.environment.getProperty(
-                    ScriptLoadBalancerProperties.TIMEOUT_PROPERTY,
-                    Long.class,
-                    DEFAULT_TIMEOUT_LENGTH
-                )
+        private final Environment environment;
+        private final MeterRegistry registry;
+        private final GenieFileTransferService fileTransferService;
+
+
+        Loader(
+            final TaskScheduler taskScheduler,
+            final Environment environment,
+            final GenieFileTransferService fileTransferService,
```

Comment:
Not to increase the scope creep here but we'll probably want to get rid of this service soon so if you were so inclined to flip this to use the `ResourceLoader` I wouldn't object